### PR TITLE
Adding Emeralds to generation

### PR DESCRIPTION
**Pull request by 12xx12**
While searching the code on learning how the server actually generates the map i noticed that emerald ore is not generated by default - now it does

https://minecraft.gamepedia.com/Emerald_Ore

### CODE REVIEW

**Comment by bearbin** [comment]
I don't like this, it feels a little too much like a special case. I think it would be better to add some configuration value to the ore types, what biomes they generate in - and then check the location when actually placing the ore nest.

**Comment by Pokechu22** [comment]
Another thing that'd need special handling (not sure if it's implemented already) is [gold ore, which generates more frequently](https://minecraft.gamepedia.com/Gold_Ore#Natural_generation) in badlands (mesa) biomes.

**Comment by 12xx12** [comment]
@Pokechu22 i just checked in with minecraft wiki and saw that too

**Comment by 12xx12** [comment]
I just added the check before to save the time of even generating anything

**Comment by bearbin** [comment]
Well, I would wait for another comment by some other developer to see what they think.

**Comment by 12xx12** [comment]
this is so a massive edge case that i would have to change the code some more to replicate vanilla behaviour because the ore is generated only in single ores but the code generates mostly two ore veines

**Comment by 12xx12** [comment]
So i made all checks pass - maybe someone will comment this.

- bearbin proposed that yould check beforehand - I don't think that is fast as you would do the setup and then dump it if the biome doesn't match. I would do the biome check before doing anything else and return if biome doesn't match

- bearbin also proposed to add a value decribing which biomes should generate - this would cause a massive list of all biomes except a few (for gold) so i would add two lists. one describing the ones allowed the other one the disallowed only one could be populated

File: src/Generating/FinishGen.cpp [hunk 1]
```diff
@@ -1763,7 +1782,16 @@ void cFinishGenOreNests::GenerateOre(
 		nestRnd /= cChunkDef::Width;
 		int BaseY = nestRnd % a_MaxHeight;
 		nestRnd /= a_MaxHeight;
-		int NestSize = a_NestSize + (nestRnd % (a_NestSize / 4));  // The actual nest size may be up to 1 / 4 larger
```
Comment (tigerw) [inline src/Generating/FinishGen.cpp:1766]: Would std::min(4, a_NestSize) work

File: Server/Plugins/APIDump/APIDesc.lua [hunk 2]
```diff
@@ -13630,6 +13630,23 @@ end
...
+							Type = "EMCSBiome",
+						}
+					},
+					Returns =
+					{
+						{
+							Type = "boolean",
+						},
+					},
+					Notes = "Returns if the biome is a mountain type biome. So mutations of the extreme hills biome"
```
Comment (tigerw) [inline Server/Plugins/APIDump/APIDesc.lua:13648]: For consistency with the others, "Returns true if the biome is mountainous (mutations of the extreme hills biome)."

File: src/BiomeDef.h [hunk 3]
```diff
@@ -151,6 +151,9 @@ extern bool IsBiomeVeryCold(EMCSBiome a_Biome);
 Doesn't report Very Cold biomes, use IsBiomeVeryCold() for those. */
 extern bool IsBiomeCold(EMCSBiome a_Biome);
 
+/** Returns true if the biome is a mountain type */
```
Comment (tigerw) [inline src/BiomeDef.h:154]: and the final full stop too.

**Comment by tigerw** [comment]
As long as there's aren't too many exceptions, it should be fine. But if Mojang gets itchy fingers (as they are wont to do...) and start adding more we'd need something more generalised. 12xx12 do you see the gold ore generation Pokechu22 mentioned as being easy to add?

File: src/Generating/FinishGen.cpp (hunk 1 above)
Comment (12xx12) [reply src/Generating/FinishGen.cpp:1766]: calculating  x mod y with 0 < y < 1 results in realy big negative values so i think
```min(a_NestSize, abs(a_NestSize + (nestRnd % (a_NestSize / 4)))```
would do the thing here

File: Server/Plugins/APIDump/APIDesc.lua (hunk 2 above)
Comment (12xx12) [reply Server/Plugins/APIDump/APIDesc.lua:13648]: done

File: src/BiomeDef.h (hunk 3 above)
Comment (12xx12) [reply src/BiomeDef.h:154]: added the new description here too

**Comment by 12xx12** [comment]
at this point it's only emeralds and the gold is some biomes

the gold should be quite simple. i'm thinking about adding another funtion biomedef.h if it's a mesa biome or not. or i change the if question to an int based one to keep the code from bloating up by doing a 20 part if test

for the gold i would do another check if the block to place is gold, sample the four spots and check if its a mesa biome and then change the values which would resolve in some magic numers in the script which i would explain

File: src/Generating/FinishGen.cpp (hunk 1 above)
Comment (tigerw) [reply src/Generating/FinishGen.cpp:1766]: I wonder if the issue is that a_NestSize / 4 does an integer division, so if it's less than 4 it rounds to zero, which blows up the modulo with a divide by zero

**Comment by tigerw** [comment]
hm. well this looks ok to me, let's see what others say

**Comment by 12xx12** [comment]
the only thing left to talk about is that with the momentarily ore generation code you generate mostly two ores next each other.

File: src/Generating/FinishGen.cpp (hunk 1 above)
Comment (12xx12) [reply src/Generating/FinishGen.cpp:1766]: so i actually rounds to zero and then crashes with a division by zero

Comment (12xx12) [reply src/Generating/FinishGen.cpp:1766]: you solution works but creates veines with at least one more as the defined size

**Comment by 12xx12** [comment]
@tigerw can you add your documentation suggestions? this was closed to fast

**Pull request merged**